vortex-array/src/array/primitive/compute/cast.rs

```diff
@@ -8,23 +8,37 @@ use crate::{Array, ArrayDType, IntoArray};
 
 impl CastFn for PrimitiveArray {
     fn cast(&self, dtype: &DType) -> VortexResult<Array> {
-        let ptype = PType::try_from(dtype)?;
-
-        // Short-cut if we can just change the nullability
-        if self.ptype() == ptype && !self.dtype().is_nullable() && dtype.is_nullable() {
-            return Ok(PrimitiveArray::new(
-                self.buffer().clone(),
-                self.ptype(),
-                Validity::AllValid,
-            )
-            .into_array());
+        let DType::Primitive(new_ptype, new_nullability) = dtype.clone() else {
```

Comment:
genuinely no clue how this gets compiled down but given nullability and ptype are copy I would drop the clone here

Reply:
they are, but DType is clone, not copy... so to have owned PType/Nullability from the destructing (instead of a ref) requires cloning the dtype

Reply:
this https://github.com/spiraldb/vortex/pull/972